File: machine-learning/small-tasks/linear-regression.cpp
#include <cstdio>
#include <algorithm>
#include <iostream>
#include <ctime>

int main() {
    std::ios_base::sync_with_stdio(false);

    int objects_amount, features_amount;
    scanf("%d %d\n", &objects_amount, &features_amount);
    int input[objects_amount][features_amount + 1];
    double x[objects_amount][features_amount + 1];

    int maxs = -1e9;
    int mins = 1e9;
    for (int i = 0; i < objects_amount; ++i) {
        for (int j = 0; j < features_amount; ++j) {
            scanf("%d", &input[i][j]);
            maxs = std::max(maxs, input[i][j]);
            mins = std::min(mins, input[i][j]);
        }
        x[i][features_amount] = 1;
        scanf("%d", &input[i][features_amount]);
    }

    int diff = maxs - mins;
    for (int i = 0; i < objects_amount; ++i) {
        for (int j = 0; j < features_amount; ++j) {
            x[i][j] = ((double) input[i][j] - mins) / diff;
        }
    }

    if (objects_amount <= 100) {
        double momentum[features_amount + 1] = {};
        double w[features_amount + 1] = {};

        for (int epoch = 0; epoch < 410000; ++epoch) {
            int i = epoch % objects_amount;

            double diff = 0;
            for (int j = 0; j < features_amount + 1; ++j) {
                diff += w[j] * x[i][j];
            }
            diff -= input[i][features_amount];

            double grad[features_amount + 1];
            for (int j = 0; j < features_amount + 1; ++j) {
                grad[j] = diff * x[i][j];
                momentum[j] = momentum[j] * 0.9 + 0.008 * grad[j];
                w[j] -= momentum[j];
            }
        }

        double s = 0.0;
        for (int j = 0; j < features_amount; ++j) {
            printf("%f\n", w[j] /= (maxs - mins));
            s += mins * w[j];
        }
        printf("%f\n", w[features_amount] - s);
        return 0;
    }


    double momentum[features_amount + 1] = {};
    double w[features_amount + 1] = {};

    for (int epoch = 0; epoch < 390000; ++epoch) {
        int i = rand() % objects_amount;

        double diff = 0;
        for (int j = 0; j < features_amount + 1; ++j) {
            diff += w[j] * x[i][j];
        }
        diff -= input[i][features_amount];

        double grad[features_amount + 1];
        for (int j = 0; j < features_amount + 1; ++j) {
            grad[j] = diff * x[i][j];
            momentum[j] = momentum[j] * 0.9 + 1e-2 * grad[j];
            w[j] -= momentum[j];
        }
    }

    double s = 0.0;
    for (int j = 0; j < features_amount; ++j) {
        printf("%f\n", w[j] /= (maxs - mins));
        s += mins * w[j];
    }
    printf("%f\n", w[features_amount] - s);

    return 0;
}

File: machine-learning/small-tasks/linear-regression.py
from sys import stdin, stdout
from random import random

def input_ints():
    return list(map(int, stdin.readline().split()))

def normalize(x):
    maxs = max([max(i[:-1]) for i in x])
    mins = min([min(i[:-1]) for i in x])

    for i in range(objects_amount):
        for j in range(features_amount):
            x[i][j] = (x[i][j] - mins) / (maxs - mins)

    return x, mins, maxs

def inverse_normalization(a, mins, maxs):
    for i in range(features_amount):
        a[i] /= maxs - mins
    a[-1] -= sum([mins * a[i] for i in range(len(a) - 1)])
    return a

def linear_regression(x, y, epochs=8000, lr=1e-2, gamma=0.9):
    x, mins, maxs = normalize(x)
    w = [0] * len(x[0])
    momentum = [0] * len(x[0])

    for epoch in range(epochs):
        i = int(random() * len(x))
        diff = sum([w[z] * x[i][z] for z in range(len(w))]) - y[i]
        shift = [diff * x[i][j] for j in range(len(x[0]))]

        momentum = [momentum[j] * gamma + lr * shift[j] for j in range(len(momentum))]
        w = [w[j] - momentum[j] for j in range(len(w))]

    return inverse_normalization(w, mins, maxs)

### main ###

# objects_amount, features_amount = input_ints()
features_amount = input_ints()[0]
objects_amount = input_ints()[0]

x = []
y = []

for i in range(objects_amount):
    line = input_ints()
    x.append(line[:-1] + [1])
    y.append(line[-1])

for i in linear_regression(x, y):
    stdout.write('{:.08f}\n'.format(i))
